python2-pybind11/challenge/containers.cpp

```diff
@@ -0,0 +1,62 @@
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include "basics.hpp"
+#include "containers.hpp"
+
+namespace py = pybind11;
+
+PYBIND11_DECLARE_HOLDER_TYPE(T, std::shared_ptr<T>);
+
+namespace containers {
+
+class DoodadSetIterator {
+public:
+    /* Keep copy of DoodadSet elements (as shared_ptrs)
+     * This is needed because the iterator might outlive the container.
+     * A more elegant solution would be to have DoodadSet inherit from
+     * std::enable_shared_from_this<DoodadSet>, so that an
+     * iterator can keep the container alive while it is still in use
+     * even though the container has been garbage collected on the Python
+     * side, but that would require modifying the C++ implementation. */
```

Comment:
From the documentation for ``return_value_policy::reference_internal``:

> Reference the object, but do not take ownership. The object is considered be owned by the C++ instance whose method or property returned it. The Python object will increase the reference count of this ‘parent’ by 1 to ensure that it won’t be deallocated while Python is using the ‘child’

Which is nice, but it isn't exactly what we want in this case. Ideally we would have Python take ownership of the returned object (the iterator) and call its destructor when it is garbage collected. But also increase the reference count of the parent by one so that doesn't get destroyed before the iterator is.
With ``reference_internal`` we would need to have ``DoodadSet`` keep track of every iterator created from it. The easiest way to do that would be to add a single iterator instance to ``DoodadSet`` itself (or a new subclass) and return const references to it instead. But that would give problems with multiple threads or even simply a double for loop over the same container (because the iterator caries state).
Therefore I think the current solution is best. Well actually I think that simply adding ``std::enable_shared_from_this`` to ``DoodadSet`` is best but that requires modifying the C++ code.
However, perhaps I can ask for a new kind of return value policy which would allow this? Or perhaps I'm missing something.